src/crimson/osd/replicated_backend.cc

```diff
@@ -142,3 +144,35 @@ seastar::future<> ReplicatedBackend::stop()
   pending_trans.clear();
   return seastar::now();
 }
+
+seastar::future<bool>
+ReplicatedBackend::already_complete(const osd_reqid_t& reqid,
+				    const eversion_t& at_version)
+{
+  auto iter = pending_trans.begin();
+  if (iter == pending_trans.end()) {
+    return seastar::make_ready_future<bool>(true);
+  }
+  auto& pending_txn = iter->second;
+  if (pending_txn.at_version > at_version) {
+    return seastar::make_ready_future<bool>(true);
+  }
+  for (; iter->second.at_version < at_version; iter++);
+  // As for now, the previous client_request with the same reqid
+  // mustn't have finished, as that would mean later client_requests
+  // has finished before earlier ones.
+  //
+  // The following line of code should be "assert(pending_txn.at_version == at_version)",
+  // as there can be only one transaction at any time in pending_trans due to
+  // PG::client_request_pg_pipeline. But there's a high possibility that we will
+  // improve the parallelism here in the future, so we loosed the restriction.
+  // Correct me if I'm wrong:-)
+  assert(iter != pending_trans.end() && iter->second.at_version == at_version);
+  if (iter->second.pending) {
+    return iter->second.all_committed.get_future().then([] {
+	  return seastar::make_ready_future<bool>(true);
+	});
+  } else {
+    return seastar::make_ready_future<bool>(true);
```

Comment:
as discussed offline, i propose following implementation
```c++
  if (std::empty(pending_trans)) {
    return seastar::now();
  }
  auto found = pending_trans.lower_bound(reqid.tid);
  assert(found != pending_trans.end());
  if (found == pending_trans.begin()) {
    // the at_version of the first pending_txn is greater than this one,
    // so the txn must have been committed
    return seastar::now();
  } else {
    // both tid and at_version are monotonically increasing, so
    // this request must be found.
    assert(found->first == reqid.tid);
    auto& pending_txn = found->second;
    if (pending_txn.pending) {
      return pending_txn.all_committed.get_future();
    } else {
      return seastar::now();
    }
  }
}
```

Reply:
@tchaikov Hi, reqid.tid is only unique and monotonic within a single session, pending_trans is indexed by another tid that exists only within the PGBackend class. So I retain the loop within the PGBackend::request_committed method, the reasons are as follows:

1. In the future, there can be scenarios like this: multiple client requests are inflight in the same PG while they are persisted to the underlying disk in the exact order of their arrival;
2. It's only when there do exists duplicated client requests, whether they are issued by the client or caused by pg interval change, that can OSDs reach this part of the code, which means this part of the code is not on the critical io execution path and the burden of the loop should be acceptable

Am I right?